### ISSUE TITLE
Replace foreignObject with standard SVG

### ISSUE THREAD

**Issue by rgoubet** [mermaid-js/mermaid#2688]
**Is your feature request related to a problem? Please describe.**
An SVG file generated by Mermaid cannot be edited using Inkscape (all boxes are black). When inserting an SGV file to an Office document, text does not appear.

**Describe the solution you'd like**
Instead of using HTML code embedded in `foreignObject` elements, Mermaid should instead use simple `text` elements (similar to Graphviz) in order to ensure that the file can be processed by SVG-compatible any application.

**Describe alternatives you've considered**
An inkscape plugin exists to convert `foreignObject` elements, but it has approximate results.

**Additional context**
See also [here](https://gitlab.com/inkscape/inbox/-/issues/3268).

**Comment by aloisklink** [comment]
Some diagrams have a `htmlLabels: false` option that you can use to remove `<foreignObject>` and use only SVG tags, see https://github.com/mermaid-js/mermaid/issues/58.

> An SVG file generated by Mermaid cannot be edited using Inkscape (all boxes are black). When inserting an SGV file to an Office document, text does not appear.

I'm pretty sure this is also because of CSS issues, see https://github.com/mermaid-js/mermaid/issues/2102, since mermaid is using some newer CSS features that are not yet supported by Inkscape. Using simpler CSS might fix it.

**Comment by rgoubet** [comment]
You're right, using this config file generates better results:

```
{
    "flowchart": {
        "htmlLabels": "false"
    }
}
```

Office will display it correctly. In Inkscape, it will only after you open the style selector (probably a bug).

**Comment by aloisklink** [comment]
It might be possible to make `htmlLabels` the default setting. Apparently `dagre-d3` had some issues with SVG styling, hence why they were using html labels, but that was back in 2015, it might be fixed now https://github.com/mermaid-js/mermaid/issues/58#issuecomment-73567103.

It's been reported too in https://github.com/mermaid-js/mermaid-cli/issues/103 and https://gitlab.com/inkscape/inbox/-/issues/3268#note_378199615, so there's quite a big demand for it.

**Comment by DaveJarvis** [comment]
Relates to (duplicate of):

* https://github.com/mermaid-js/mermaid-cli/issues/103
* https://github.com/mermaid-js/mermaid/issues/2102
* https://github.com/mermaid-js/mermaid/issues/2485

**Comment by lmoellendorf** [comment]
## Workaround

Generate a PDF:
```
~/node_modules/.bin/mmdc -i images/block/sync.mermaid -o ./images/block/sync.pdf
```
Use [pdf2svg](https://github.com/dawbarton/pdf2svg/) to extract SVG:
```
pdf2svg images/block/sync.pdf images/block/sync.0.svg
```
Now use inkscape to resize page to content (as in Inkscape and in _Document Properties_ => _Resize to content_):
```
inkscape --export-area-drawing --export-filename=images/block/sync.svg images/block/sync.0.svg
```

**Comment by Foadsf** [comment]
@lmoellendorf why not using the `--pdfFit` option? 🤔

**Comment by jongalloway** [comment]
I ended up running a script in browser tools to convert foreignObjects and HTML to text. I'm probably missing some elements since I was using a state diagram, but this might be a start for someone else hitting this: https://gist.github.com/jongalloway/ce88bdde32d01bd94ccb13716a66d271

**Comment by Marcono1234** [comment]
@jgreywolf, was this actually solved and if so which commit or pull request fixed this? It seems there are still a few usages of `foreignObject`.

Just wanted to ask to be sure this issue was not closed by accident.

**Comment by jgreywolf** [comment]
This was closed as it is duplicate of many other issues - specifically that mermaid uses a browser based svg renderer with a lot of CSS tweaks which lead to issues in other applications, as well as the inclusion of all of these foreign objects.

The following are the ones that are still open:
https://github.com/mermaid-js/mermaid/issues/2169
https://github.com/mermaid-js/mermaid/issues/1923
https://github.com/mermaid-js/mermaid/issues/1845

If you feel that this would not be resolved with one of those, or that it really needs to be its own issue, please feel free to reopen it

**Comment by jlpoolen** [comment]
I modified Jon's script for use in a console since Inkscape 1.4.3 (0d15f75042, 2025-12-25) still creates foreignObjects.  See https://gist.github.com/jongalloway/ce88bdde32d01bd94ccb13716a66d271?permalink_comment_id=6198541#gistcomment-6198541
